src/contracts/GPv2Settlement.sol

```diff
@@ -196,12 +196,14 @@ contract GPv2Settlement {
         uint256[] calldata clearingPrices,
         bytes calldata encodedTrades
     ) internal returns (GPv2TradeExecution.Data[] memory executedTrades) {
-        uint256 tradeCount = encodedTrades.tradeCount();
+        (uint256 tradeCount, bytes calldata remainingEncodedTrades) =
+            encodedTrades.decodeTradeCount();
         executedTrades = new GPv2TradeExecution.Data[](tradeCount);
 
         GPv2Encoding.Trade memory trade;
-        for (uint256 i = 0; i < tradeCount; i++) {
-            encodedTrades.tradeAtIndex(i).decodeTrade(
+        uint256 i = 0;
+        while (remainingEncodedTrades.length != 0) {
```

Comment:
Just curious, any reason why you chose a while loop instead of a `for` loop?
```
for (uint256 i = 0; i < tradeCount; i++) {
    // ...
}
require(remainingEncodedTrades.length == 0, "GPv2: invalid trade encoding");
```

Reply:
For consistency with interaction decoding, where the interaction count is not known. Will test this change and report back!

Reply:
It looks like changing it into a `for` loop makes it negligibly more inefficient (~5 gas/trade). Interestingly it doesn't appear to depend on the absolute number of orders benched.

<details><summary>bench</summary>

```
=== Settlement Gas Benchmarks ===
--------------+--------------+--------------+--------------+--------------
       tokens |       trades | interactions |      refunds |          gas 
--------------+--------------+--------------+--------------+--------------
            2 |           10 |            0 |            0 |       741864  (change: 5 / trade)
            3 |           10 |            0 |            0 |       742480  (change: 5 / trade)
            4 |           10 |            0 |            0 |       751424  (change: 8 / trade)
            5 |           10 |            0 |            0 |       756120  (change: 3 / trade)
            6 |           10 |            0 |            0 |       752404  (change: 2 / trade)
            7 |           10 |            0 |            0 |       761360  (change: 11 / trade)
            8 |           10 |            0 |            0 |       766044  (change: 5 / trade)
            8 |           20 |            0 |            0 |      1453196  (change: 3 / trade)
            8 |           30 |            0 |            0 |      2094012  (change: 7 / trade)
            8 |           40 |            0 |            0 |      2739290  (change: 6 / trade)
            8 |           50 |            0 |            0 |      3384193  (change: 6 / trade)
            8 |           60 |            0 |            0 |      4025704  (change: 5 / trade)
            8 |           70 |            0 |            0 |      4671599  (change: 7 / trade)
            8 |           80 |            0 |            0 |      5317304  (change: 6 / trade)
            2 |           10 |            1 |            0 |       812454  (change: 6 / trade)
            2 |           10 |            2 |            0 |       858200  (change: 8 / trade)
            2 |           10 |            3 |            0 |       903923  (change: 5 / trade)
            2 |           10 |            4 |            0 |       949555  (change: 6 / trade)
            2 |           10 |            5 |            0 |       995363  (change: 8 / trade)
            2 |           10 |            6 |            0 |      1041065  (change: 6 / trade)
            2 |           10 |            7 |            0 |      1086826  (change: 2 / trade)
            2 |           10 |            8 |            0 |      1132461  (change: 3 / trade)
            2 |           50 |            0 |           10 |      3109915  (change: 7 / trade)
            2 |           50 |            0 |           15 |      3066599  (change: 7 / trade)
            2 |           50 |            0 |           20 |      3023107  (change: 4 / trade)
            2 |           50 |            0 |           25 |      2979683  (change: 2 / trade)
            2 |           50 |            0 |           30 |      2936487  (change: 5 / trade)
            2 |           50 |            0 |           35 |      2893111  (change: 5 / trade)
            2 |           50 |            0 |           40 |      2849679  (change: 5 / trade)
            2 |           50 |            0 |           45 |      2806399  (change: 7 / trade)
            2 |           50 |            0 |           50 |      2763023  (change: 4 / trade)
            2 |            2 |            0 |            0 |       182302  (change: 1 / trade)
            2 |            1 |            1 |            0 |       182899  (change: 9 / trade)
           10 |          100 |           10 |           20 |      7115880  (change: 6 / trade)
```
</details>

Reply:
Kind of funny :stuck_out_tongue:. I guess we should keep the while loop then.